runtime/scripts/results.js

```diff
@@ -0,0 +1,211 @@
+                        header.push(questionName + " " + partName + " " + stepName + " Marks");
+                        trueOrder.push(trueQuestionNumber + " " + partName + " " + stepName + " Marks");
+                        expectedAnswers.push(stepObject.marks);
+                        studentAnswers.push(stepObject.score);
+                        header.push(questionName + " " + partName + " " + stepName + " Answer");
+                        trueOrder.push(trueQuestionNumber + " " + partName + " " + stepName + " Answer");
+                        expectedAnswers.push(stepObject.getCorrectAnswer(stepObject.getScope()));
+                        studentAnswers.push(stepObject.studentAnswer);
+                    }
+                }
+            }
+            let dataset = [header, expectedAnswers, studentAnswers];
+            if (randomised) {
+                dataset = [header, trueOrder, expectedAnswers, studentAnswers];
+            }
+            return Numbas.results.make_csv(dataset);
+        },
+
+        /**
+         * 
+         * @param {string} file 
+         */
+        create_and_download_file : function(file) {
+            //pulled from https://stackoverflow.com/questions/8310657/how-to-create-a-dynamic-file-link-for-download-in-javascript
+            let mime_type = 'text/plain';
+            var blob = new Blob([file],{type: mime_type});
+            var dlink = document.createElement('a');
+            document.body.appendChild(dlink); //may be necessary for firefox/some browsers
+            dlink.download="results.csv";
+            dlink.href = window.URL.createObjectURL(blob);
+            dlink.onclick = function(e) {
+              var that = this;
+              setTimeout(function() {
+                window.URL.revokeObjectURL(that.href);
+              }, 1500);
+            };
+          
+            dlink.click()
+            dlink.remove()
+            
+          },
+        /**
+         * Checks if any aspect of the exam is in a randomised order
+         * 
+         * @returns {boolean} whether the exam has any randomised components
+         */
+        examIsRandomised : function() {
+            let exam = Numbas.exam;
+            if (exam.settings.shuffleQuestionGroups) {
+                return true;
+            }
+            for (let groupKey of Object.keys(exam.question_groups)) {
+                let group = exam.question_groups[groupKey];
+                if (group.settings.pickingStrategy!="all-ordered"){ //Future-proof: assume any new picking strategy might be randomising
+                    return true;
+                }
+            }
+            return false;
+
+        },
+
+        /**
+         * Provides the mapping between the questions as shown to the student and the editor-order of those questions
+         * 
+         * @returns {Array.<int>} the mapping of the questions as shown to the student to the editor order.
+         */
+        questionMapping : function() {

```

Comment:
I think it makes more sense to maintain the question group structure. When you have a "take N questions" group, I think it would be helpful to see which questions come from that group.
So another two rows above the question names row could give the group names, and then the initial position of each question in that group.